### PR TITLE
Add EnumConfParser to SparkConfParser

### PR DESCRIPTION
**Pull request by huaxingao**
Add `EnumConfParser` to `SparkConfParser` to parse Enum type properties

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @aokolnychyi

**Review by nastra — APPROVED** [review]

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java [hunk 1]
```diff
@@ -197,6 +201,34 @@ private Duration toDuration(String time) {
     }
   }
 
+  class EnumConfParser<T extends Enum<T>> extends ConfParser<EnumConfParser<T>, T> {
+    private T defaultValue;
+    private final Function<String, T> toEnum;
```
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:206]: Minor: We usually have final vars before mutable. Can we switch the order of these two?

**Review by aokolnychyi — APPROVED** [review]

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java [hunk 2]
```diff
@@ -70,6 +70,10 @@ public DurationConfParser durationConf() {
     return new DurationConfParser();
   }
 
+  public <T extends Enum<T>> EnumConfParser<T> enumConf(Function<String, T> func) {
```
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:73]: Minor: What about `func` -> `toEnum`?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkReadConf.java [hunk 3]
```diff
@@ -302,14 +302,12 @@ public PlanningMode dataPlanningMode() {
...
-            .sessionConf(SparkSQLProperties.DATA_PLANNING_MODE)
-            .tableProperty(TableProperties.DATA_PLANNING_MODE)
-            .defaultValue(TableProperties.PLANNING_MODE_DEFAULT)
-            .parse();
-    return PlanningMode.fromName(modeName);
+    return confParser
+        .enumConf(PlanningMode::fromName)
+        .sessionConf(SparkSQLProperties.DATA_PLANNING_MODE)
+        .tableProperty(TableProperties.DATA_PLANNING_MODE)
+        .defaultValue(PlanningMode.fromName(TableProperties.PLANNING_MODE_DEFAULT))
```
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkReadConf.java:309]: We can also offer `defaultValue(String value)` and call `toEnum(value)` internally.

**Review by aokolnychyi — APPROVED** [review]
LGTM, pending tests.

**Comment by huaxingao** [comment]
Thanks @aokolnychyi @nastra

**Pull request merged**